File: 3116-kth-smallest-amount-with-single-denomination-combination/3116-kth-smallest-amount-with-single-denomination-combination.java
class Solution {

    public long gcd(long a, long b) {
        while (b != 0) {
            long temp = a % b;
            a = b;
            b = temp;
        }
        return a;
    }

    public long lcm(long a, long b) {
        return a / gcd(a, b) * b;
    }

    // Count how many different amounts <= x
    // can be made using at least one coin.
    public long count(long x, int[] coins) {

        int n = coins.length;
        long total = 0;

        // Try every non-empty subset of coins
        for (int mask = 1; mask < (1 << n); mask++) {

            long multiple = 1;
            int bits = 0;

            for (int i = 0; i < n; i++) {

                if ((mask & (1 << i)) != 0) {

                    bits++;

                    multiple = lcm(multiple, coins[i]);

                    // No multiple of this LCM can be <= x
                    if (multiple > x) {
                        break;
                    }
                }
            }

            if (multiple > x) {
                continue;
            }

            long ways = x / multiple;

            // Odd subset -> add
            if (bits % 2 == 1) {
                total += ways;
            }
            // Even subset -> subtract
            else {
                total -= ways;
            }
        }

        return total;
    }

    public long findKthSmallest(int[] coins, int k) {

        long left = 1;

        // Maximum possible answer:
        // smallest coin * k
        long right = (long) coins[0] * k;

        for (int coin : coins) {
            right = Math.min(right, (long) coin * k);
        }

        // Binary search
        while (left < right) {

            long mid = left + (right - left) / 2;

            if (count(mid, coins) >= k) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }

        return left;
    }
}

File: 3116-kth-smallest-amount-with-single-denomination-combination/3116-kth-smallest-amount-with-single-denomination-combination.py
class Solution(object):

    def findKthSmallest(self, coins, k):
        """
        :type coins: List[int]
        :type k: int
        :rtype: int
        """

        def gcd(a, b):
            while b:
                a, b = b, a % b
            return a

        def lcm(a, b):
            return a // gcd(a, b) * b

        # Count valid amounts <= x
        def count(x):

            n = len(coins)
            total = 0

            # Try every non-empty subset
            for mask in range(1, 1 << n):

                multiple = 1
                bits = 0
                valid = True

                for i in range(n):

                    if mask & (1 << i):

                        bits += 1

                        multiple = lcm(multiple, coins[i])

                        if multiple > x:
                            valid = False
                            break

                if not valid:
                    continue

                ways = x // multiple

                # Odd number of coins -> add
                if bits % 2 == 1:
                    total += ways

                # Even number of coins -> subtract
                else:
                    total -= ways

            return total

        # Binary search
        left = 1
        right = min(coins) * k

        while left < right:

            mid = left + (right - left) // 2

            if count(mid) >= k:
                right = mid
            else:
                left = mid + 1

        return left
        